File: src/main/java/at/yawk/math/algorithm/GcdSolver.java
package at.yawk.math.algorithm;

import at.yawk.math.data.Expressions;
import at.yawk.math.data.IntegerExpression;
import lombok.experimental.UtilityClass;

/**
 * @author yawkat
 */
@UtilityClass
public class GcdSolver {
    public static IntegerExpression gcd(IntegerExpression a, IntegerExpression b) {
        if (!a.isPositive()) { throw new IllegalArgumentException(a + " is not positive"); }
        if (!b.isPositive()) { throw new IllegalArgumentException(b + " is not positive"); }
        return Expressions.integer(a.getValue().gcd(b.getValue()));
    }
}

File: src/main/kotlin/at/yawk/math/algorithm/GcdSolver.kt
package at.yawk.math.algorithm

import at.yawk.math.data.Expression
import at.yawk.math.data.Expressions
import at.yawk.math.data.IntegerExpression
import at.yawk.math.data.Sign

/**
 * @author yawkat
 */
object GcdSolver {
    fun gcd(a: IntegerExpression, b: IntegerExpression): Expression {
        if (a.sign != Sign.POSITIVE) {
            throw IllegalArgumentException("$a is not positive")
        }
        if (b.sign != Sign.POSITIVE) {
            throw IllegalArgumentException("$b is not positive")
        }
        return Expressions.int(a.value.gcd(b.value))
    }
}